### PR TITLE
Adding multiple functionality support to diarization recipe

### PR DESCRIPTION
**Pull request by nauman-daw**
Hi @mravanelli ,
This PR will add functionalities like multi-microphone array beamforming, distant microphones, and different backends.

### CODE REVIEW

**Comment by mravanelli** [comment]
Hi @nauman-daw,
thank you for this PR.  I took a first look at the new code and these are my comments:

1. `csv-to-json`: This seems to be quite unnatural. What about if we fully revise the data_prep script and we can directly create a JSON file that works with both single and multiple microphones?

2. I think these lines in  https://github.com/speechbrain/speechbrain/pull/642/files#L47 are not optimal: 
```
        fs = 16000  # see if this can be automatically adjusted.
        stft = STFT(sample_rate=fs)
        cov = Covariance()
        gccphat = GccPhat()
        delaysum = DelaySum()
        istft = ISTFT(sample_rate=fs)
```
Basically, you initialize the classes for every new sentence read from the disk. This might be very slow. These classes should be initialized only once. This looks like a code snippet that many users might want to use. What about if we create a class in lobes/multi-mic that performs delay-and-sum beamforming given a multi-channel input?

3. The threshold tuning part is a bit hard to follow (especially for people not 100% expert in diarization). My suggestion is to describe it better in the docstring of experiment.py

4. Make sure the README file reports exactly the performance of the paper.

**Comment by nauman-daw** [comment]
Hi @mravanelli ,

Thank you for the reviews. Please find my response below.

> Hi @nauman-daw,
> thank you for this PR. I took a first look at the new code and these are my comments:
> 
> 1. `csv-to-json`: This seems to be quite unnatural. What about if we fully revise the data_prep script and we can directly create a JSON file that works with both single and multiple microphones?

Sure. That would be a cleaner approach. It may take a carefully updating data_prep pipeline, as I have to take subsets from the complete set of recordings. This might take some time. Wouldn't it be nice to have a separate PR for that? 

> 2. I think these lines in  https://github.com/speechbrain/speechbrain/pull/642/files#L47 are not optimal:
> 
> ```
>         fs = 16000  # see if this can be automatically adjusted.
>         stft = STFT(sample_rate=fs)
>         cov = Covariance()
>         gccphat = GccPhat()
>         delaysum = DelaySum()
>         istft = ISTFT(sample_rate=fs)
> ```
> 
> Basically, you initialize the classes for every new sentence read from the disk. This might be very slow. These classes should be initialized only once. This looks like a code snippet that many users might want to use. What about if we create a class in lobes/multi-mic that performs delay-and-sum beamforming given a multi-channel input?
> 

Sure. I first shifted this outside. But now we a separate class in lobes/ . Note that, the beamforming is slow compared to other audio stream. This is expected as we have almost 8 times more data.

> 1. The threshold tuning part is a bit hard to follow (especially for people not 100% expert in diarization). My suggestion is to describe it better in the docstring of experiment.py

 It would be easier to follow if we start reading from main(). However, I added a few more comments that may make it a bit easier to read. Please let me know if you feel we should still add more comments. 

> 2. Make sure the README file reports exactly the performance of the paper.

Updated README file

**Comment by mravanelli** [comment]
Hi @nauman-daw,
I can review this PR now. However, there are some conflicts to solve. Could you please take a look at that?

**Comment by mravanelli** [comment]
Hi @nauman-daw , 
thank you for this PR and sorry for my delay with the review. I left some minor comments above. Moreover, I have the following ones:
1. Some small conflicts need to be solved.
2. `csv_to_json` => My suggestion is to take this opportunity to re-write the data preparation and put it in JSON from the beginning (everything will be much cleaner in this way). This way, we don't need this hacky conversion. However, the function  `csv_to_json `looks general enough and I suggest putting it in `speechbrain/utils/data_utils.py` (with a proper docstring and an example).
3. diarize_dataset => This is crucial function. Please add more comments + a more detailed docstring.
4. `dataio_prep_multi_mic`, `dataio_prep` => there are unneeded new lines here.
5. `check_dirs() `=> I think the first two checks should be done in prepare_ami.py directly. As far as I remember, the last check is not needed because done in create_experiment_directory
6. The final part of the code evaluates diarization on both dev and test. The code is basically the same for both so I'm wondering if we can save about 50% of the code just by using a for loop here like "for set in ['dev','test']:"
7. The part about setting the threshold takes a lot of space. If you think this is general enough, I think we can put it in `processing/diarization.py`, right? This way we might describe the functions better with longer docstrings. It looks also that there is a significant overlap in the code between the functions (e.g, `dev_p_tuner`, `dev_p_tuner`, `dev_nn_tuner`) and maybe there is a way to save some code here as well.

File: README.md [hunk 1]
```diff
@@ -70,7 +70,7 @@ The recipes released with speechbrain implement speech processing systems with c
 | TIMIT      | Speech Recognition | wav2vec2 + CTC/Att. | PER=8.04% (test) |
 | CommonVoice (French) | Speech Recognition | wav2vec2 + CTC/Att. | WER=13.7% (test) |
 | VoxCeleb2      | Speaker Verification | ECAPA-TDNN | EER=0.69% (vox1-test) |
-| AMI      | Speaker Diarization | ECAPA-TDNN | DER=2.13% (lapel-mix)|
+| AMI      | Speaker Diarization | ECAPA-TDNN | DER=2.57% (LapelMix-eval)|
```
[the remark this answers is not in the document]
Comment (nauman-daw) [reply README.md:73]: Yes. It will be better. updated

File: speechbrain/lobes/beamform_multimic.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+from speechbrain.processing.multi_mic import (
+    Covariance,
+    GccPhat,
+    DelaySum,
+)
+
+
+class DelaySum_Beamformer(torch.nn.Module):
+    """Generate features for input to the speech pipeline.
```
[the remark this answers is not in the document]
Comment (nauman-daw) [reply speechbrain/lobes/beamform_multimic.py:20]: Thanks for pointing this out. Corrected the docstring now.

**Comment by mravanelli** [comment]
Thank you  @nauman-daw! Excellent work!
